### DISCUSSION TITLE
reproviding the remote service history

### DISCUSSION THREAD

**Discussion by demel42** [bimmerconnected/bimmer_connected#383] [Feature request] [answered]
Hi,

thanks for your work on new api. i am using your work, to get data into the smarthome "IP-Symcon" (a solution in germany)

i have a question:
in the old api, there was a call to get set serviceHistory, so take al look on the remote services

with the new api, the old call it disfunctional - as expected.

in the vehicle-data is a item _isRemoteHistorySupported_ an in the original BMW-app there ist a item "Remote-Verlauf" (in Germany), would be call "remote history" - in the App, i can see the data.
My problem is my lack of ability to debug the app communication (since it is https).
Is there any idea what the API call could be?

Thanks a lot

**Comment by rikroe** [comment, marked as answer]
Hi demel,

thats actually on my todo list, however I am not fully certain how I want to implement it (i.e. if & where to store it on a vehicle object).
I'm currently not quite happy how some attributes are accessed from the  `vehicle`, some from `vehicle status` (or directly from the vehicle), etc., so I am thinking about taking a step back and fully-designing the library.

The URL called is basically `/eadrax-vrccs/v2/presentation/remote-history/{vin}`. How are you accessing the library? 
Via CLI or from Python?

With the latter, you could implement a workaround like  blow (not tested ;) ):
```python
for vehicle in account.vehicles:
  account.send_request(
    BASE_URL + "/eadrax-vrccs/v2/presentation/remote-history/{vin}".format(server=account.server_url, vin=vehicle.vin),
    brand=vehicle.brand
  )
```

**Comment by demel42** [comment]
Hi rikoe,

thanks a lot, it works.
I didn't use the library or cli, Modules in IP-Symcon are written in PHP and because of the architecture it is easier to implement modules without using system extensions.
The BMW module used to be based on the named information from https://github.com/edent/BMW-i-Remote and others as well as snooping with wireshark, but since the changes a long time ago I had to adapt that too and then came across your great work.
I take your ideas and adapt them in PHP - but due to the consistent encryption via https I can no longer get the necessary information and am therefore dependent on help in the case of the BMW module.

**Comment by demel42** [comment]
Hi rikoe,

your Information was a good help.
are there any api-calls, you know and didn't use yet?

in the old api, there where information about the battery (battery_size_max, socMax)an some efficiency hint's and info about the last trips.

any additional information would be a great help

**Comment by rikroe** [comment]
Unfortunately, I haven't seen any other API endpoints that have socMax available.

There was a discussion that a login against the old API's `authorize` could still work for old `navigation` endpoint containing socMax, but I cannot find the responding PR or discussion item.

On my tries I at least was not able to get any data back (403 forbidden response from server).

**Reply by demel42** [reply]
Hi,
i have just tried ist again. the old Authentication succeeds but the API-call `/api/vehicle/navigation/v1`and `/api/vehicle/dynamic/v1/`  returns a HTTP 301(Moved Permanently) - the behaviour i had seen the last weeks.
i also tried some API calls in the last days which sounded promising to me, but yes, there is only a 403 on it or a 500(Server-Error) 

but what the heck, most of the data is now back thanks to your help.






I am still investigating the behavior of chargingState and charging_end_time at the moment.

i have a forum member who only ever loads the car at night ...

```
[chargingProfile] => Array
     (
         [reductionOfChargeCurrent] => Array
             (
                 [start] => Array
                     (
                         [hour] => 2
                         [minute] => 0
                     )
                 [end] => Array
                     (
                         [hour] => 6
                         [minute] => 55
                     )
                )
        [chargingMode] => delayedCharging
        [chargingPreference] => chargingWindow
        [chargingControlType] => weeklyPlanner
```

... wait for charging to begin

```
[chargingState] => Array
    (
        [chargePercentage] => 12
        [state] => CHARGING
        [type] => CONDUCTIVE 
        [isChargerConnected] => 1
    )                   

[infoLabel] => Startet um ~ 02:00 AM
[chargingType] =>
[chargingStatusType] => PLUGGED_IN
[chargingStatusIndicatorType] => PLUGGED_IN
```

... really charging

```
[chargingState] => Array
    (
        [chargePercentage] => 45
        [state] => CHARGING
        [type] => CONDUCTIVE
        [isChargerConnected] => 1
    )

[infoLabel] => Laden
[chargingType] => charging
[chargingStatusType] => CHARGING
[chargingStatusIndicatorType] => CHARGING

```

When _chargingPreference_ has the value _chargingWindow_  or potentially _chargingMode_ is _delayedCharging_  the BMW API misleadingly outputs _chargingState.state_ as _CHARGING_.
in reality this is rather READY_TO_LOAD or WAIT_TO_START.
Only when _chargingStatusIndicatorType_ changes to _CHARGING_ it is really charging.
Furthermore, in this case the end time is not output in _infoLabel_.

I will report when I have found out something reliable.

**Comment by rikroe** [comment]
Yes, I have heard that regarding charging/just plugged in as well. 
I am trying to base my logic only on `fuelIndicators` as they seem to be the only ones showing correct values (see https://github.com/bimmerconnected/bimmer_connected/pull/385).

If your car supports all that fancy new stuff and you have an rooted android phone lying around, I can help setting you up to be able to to MITM against the current app again.

**Reply by demel42** [reply]
at this moment, i have only an X5 from 2015, with very reduced digital information (a little bit more than  the remote services) and a X3 from 2019 with more informations. but 2 month ago i ordered a ix3 with planned delivery end of Q1/2022.

and i'm very interesed in smarthome, IoT and such things; if you take a look at my github-repository you'l see my interests and bmw is one of it.

i'm Apple-minded and have only iPhones, but it will be possible to get an andriod mobile phone. what's the requirement of such an android device?

**Reply by rikroe** [reply]
It works with iPhones as well, you just need a jailbreak.

I'll try to write together as much as possible over the holidays. As I'm fully android, you then just have to translate where needed. But it 100% works on with iPhones as well.

**Reply by demel42** [reply]
will be no problem, to get an old Android. i think used devices are available for little money. or i ask around with colleagues or friends..

have you any requirements (android-version an so on)?

**Reply by rikroe** [reply]
MyBMW requires at least Android 6.0. Frida (the tool required for disabling the SSL inspection recommends using Android 9, but older versions should work as well.
The device I am using is running Android 7.1.2.

Just make sure to get a phone that is either rooted or can be rooted (best check the [XDA forums](https://forum.xda-developers.com/) for that).

**Reply by rikroe** [reply]
Added some documentation in #392. Either you wait until its merged & part of readthedocs or skim through it now :)

EDIT: it's merged, please see [here](https://bimmer-connected.readthedocs.io/en/latest/development/reverse_engineering_mybmw.html). Feedback is appreciated!
